### PR TITLE
Separate sponsor recipient

### PR DESCRIPTION
**Pull request by forshtat**

### CODE REVIEW

File: test/relay_client_test.js [hunk 1]
```diff
@@ -146,7 +150,8 @@ contract('RelayClient', function (accounts) {
         from: gasLess,
         to: to,
         txfee: 12,
-        gas_limit: 1000000
+        gas_limit: 1000000,
```
Comment (drortirosh) [inline test/relay_client_test.js]: I saw elsewhere that you changed it to gasLimit

File: test/relay_client_test.js [hunk 2]
```diff
@@ -582,30 +596,32 @@ contract('RelayClient', function (accounts) {
...
+        await sr.emitMessage('hello world', { from: gasLess, gasSponsor: gasSponsor2.address })
         assert.fail()
       } catch (error) {
-        assert.equal(true, error.message.includes('The relay hub address is set to zero in recipient at'))
+        assert.equal(true, error.message.includes('The relay hub address is set to zero in sponsor at'))
       }
     })
 
     it('should throw on invalid recipient', async function () {
       const tbk = new RelayClient(web3)
```
Comment (drortirosh) [inline test/relay_client_test.js]: haha. didn't know we still have tbk here..

File: .gitignore [hunk 3]
```diff
@@ -17,3 +17,4 @@
 /singleton/
 /server/pkg/
 /.eslintcache
+/src/js/relayclient/interfaces/
```
[the remark this answers is not in the document]
Comment (forshtat) [reply .gitignore]: was there before. will change in a separate commit

File: contracts/BaseGasSponsor.sol [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+import "./interfaces/IRelayHub.sol";
+
+/**
+ * Abstract base class to be inherited by a concrete Gas Sponsor
+ * A subclass must implement:
+ *  - acceptRelayCall
+ *  - preRelayedCall
+ *  - postRelayedCall
+ */
+contract BaseGasSponsor is IGasSponsor, Ownable {
```
[the remark this answers is not in the document]
Comment (forshtat) [reply contracts/BaseGasSponsor.sol]: I do not exactly agree with GasPayer as a name

[the remark this answers is not in the document]
Comment (forshtat) [reply contracts/BaseGasSponsor.sol]: what about 'withdraw'? Shouldn't base gas sponsor be able to withdraw? If needed, TokenPayer can implement IGasSponsor directly

File: contracts/BaseGasSponsor.sol [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    }
+    /*
+     * modifier to be used by recipients as access control protection for preRelayedCall & postRelayedCall
+     */
+    modifier relayHubOnly() {
+        require(msg.sender == getHubAddr(), "Function can only be called by RelayHub");
+        _;
+    }
+
+    function setRelayHub(IRelayHub hub) public onlyOwner {
```
[the remark this answers is not in the document]
Comment (forshtat) [reply contracts/BaseGasSponsor.sol]: will fix
###2

File: contracts/RelayHub.sol [hunk 6]
```diff
@@ -312,11 +312,11 @@ contract RelayHub is IRelayHub {
 
         // We've already checked that the recipient has enough balance to pay for the relayed transaction, this is only
         // a sanity check to prevent overflows in case of bugs.
-        require(balances[relayRequest.callData.target] >= charge, "Should not get here");
-        balances[relayRequest.callData.target] -= charge;
+        require(balances[relayRequest.relayData.gasSponsor] >= charge, "Should not get here");
+        balances[relayRequest.relayData.gasSponsor] -= charge;
```
[the remark this answers is not in the document]
Comment (forshtat) [reply contracts/RelayHub.sol]: Will perform an overall refactor of relayhub, as there is a lot of code that can be simplified now that we are not constantly on 16 variables edge
###3

File: server/src/librelay/relay_server_test.go [hunk 7]
```diff
@@ -384,7 +406,7 @@ func assertNoTransactionResent(t *testing.T, relay *RelayServer) {
 }
 
 func TestCreateRelayTransaction(t *testing.T) {
-	request := newRelayTransactionRequest(t, 0, "0xd735f61c4364b4543fe627529959558ea14f7d50940347c2073db1ef1d18691958be8c4da835d65bb84e74eff73c428085dc1b3fce14b34fb024e27c4e6921391c")
+	request := newRelayTransactionRequest(t, 0, "0xbe894dfa52ed84c3395d609a6976177f74b75748e57ffefc25c9b4dc134e36b37f7621f297cd465be9c2ede807e91791a6d57b34e1fefaf5e19a39c535d20bbf1c")
```
[the remark this answers is not in the document]
Comment (forshtat) [reply server/src/librelay/relay_server_test.go]: hqdefault

[the remark this answers is not in the document]
Comment (forshtat) [reply server/src/librelay/relay_server_test.go]: will be deprecated with the entire golang stack soon, so don't bother

File: src/js/relayclient/EIP712/RelayData.js [hunk 8]
```diff
@@ -1,9 +1,10 @@
 class RelayData {
-  constructor ({ senderAccount, senderNonce, relayAddress, pctRelayFee }) {
+  constructor ({ senderAccount, senderNonce, relayAddress, pctRelayFee, gasSponsor }) {
```
[the remark this answers is not in the document]
Comment (forshtat) [reply src/js/relayclient/EIP712/RelayData.js]: I have an outstanding task of bringing order and consistency to the function params.
###4

File: src/js/relayclient/utils.js [hunk 9]
```diff
@@ -84,7 +87,7 @@ module.exports = {
     })
   },
 
-  getRelayRequest: function (sender, recipient, txData, fee, gasPrice, gasLimit, senderNonce, relay) {
+  getRelayRequest: function (sender, recipient, txData, fee, gasPrice, gasLimit, senderNonce, relay, gasSponsor) {
```
[the remark this answers is not in the document]
Comment (forshtat) [reply src/js/relayclient/utils.js]: ###4 as well

File: test/relay_client_test.js (hunk 1 above)
Comment (forshtat) [reply test/relay_client_test.js]: ###4 as well

File: test/relay_client_test.js (hunk 2 above)
Comment (forshtat) [reply test/relay_client_test.js]: Will get rid of 'tbk' names of variables
###5

**Pull request merged**